File: python/src/inillucent/_check.py
"""Turning a failed C call into the right Python exception."""

from __future__ import annotations

from ._library import lib
from .errors import UNSUPPORTED, InillucentError, UnsupportedError, status_name
from .rows import decode


def raise_from_error(handle) -> None:
    """Build the exception a C error describes, and free the error either way.

    The free happens in a finally because an exception built from an error must
    not leak it.

    @param handle - the C error handle, which is owned by this call
    """
    library = lib()
    try:
        status = library.inillucent_error_status(handle)
        message = decode(library.inillucent_error_message(handle)) or ""
        feature = decode(library.inillucent_error_feature(handle))
        detail = decode(library.inillucent_error_detail(handle))
        offset = library.inillucent_error_offset(handle)
    finally:
        library.inillucent_error_free(handle)
    kind = UnsupportedError if status == UNSUPPORTED else InillucentError
    raise kind(status, message, feature, detail, offset)


def check(status: int, error) -> None:
    """Raise when a call failed, using the error it produced.

    A non zero status with no error still raises: a call that failed and said
    nothing is not a reason to carry on.

    @param status - what the call returned
    @param error - the error out parameter the call was given
    """
    if status == 0:
        return
    if error:
        raise_from_error(error)
    raise InillucentError(status, f"the call failed with {status_name(status)}")

File: typescript/src/check.ts
/**
 * Turning a failed C call into the right JavaScript exception.
 */

import { calls } from './ffi.js';
import { InillucentError, Status, UnsupportedError, statusName } from './errors.js';

/**
 * Builds the exception a C error describes, and frees the error either way.
 *
 * The free happens in a finally because an exception built from an error must
 * not leak it.
 *
 * @param handle - the C error handle, which is owned by this call
 */
function errorFrom(handle: unknown): InillucentError {
  const c = calls();
  try {
    const status = c.error_status(handle) as number;
    const message = (c.error_message(handle) as string) ?? '';
    const feature = (c.error_feature(handle) as string) || undefined;
    const detail = (c.error_detail(handle) as string) || undefined;
    const reported = c.error_offset(handle) as number;
    const offset = reported >= 0 ? reported : undefined;
    const kind = status === Status.Unsupported ? UnsupportedError : InillucentError;
    return new kind(status, message, feature, detail, offset);
  } finally {
    c.error_free(handle);
  }
}

/**
 * Throws when a call failed, using the error it produced.
 *
 * A non zero status with no error still throws: a call that failed and said
 * nothing is not a reason to carry on.
 *
 * @param status - what the call returned
 * @param error - the error out parameter the call was given
 */
export function check(status: number, error: [unknown]): void {
  if (status === Status.Ok) return;
  if (error[0]) throw errorFrom(error[0]);
  throw new InillucentError(status, `the call failed with ${statusName(status)}`);
}
